### DISCUSSION TITLE
Encoder counter mode

### DISCUSSION THREAD

**Discussion by horcsab** [Peter-van-Tol/LiteX-CNC#109] [General]
Dear Peter,

Could you add the counter mode to the encoder input?
I would like to measure voltage with U/f (LM311) converter and would like to use it for that.
Or do you have an idea how this could be solved in the simplest way?
 I want to process 0-10V with 8-16bit resolution.

**Comment by Peter-van-Tol** [comment]
I think there are 3 ways to solve what you want:
- the only one possible at this moment: Adc  conversion to parallel interface. This uses 8 to 16 inputs of your FPGA and you have to reassemble the individual bits to a value in your HAL file. The only advantage of this system is that it is possible with the current firmware.
- using a Adc to serial chip. These can be picked up for around 10 dollars on Aliexpress. SPI nor I2C is supported yet by LitexCNC, but it is my wish to support this in near future. Biggest hudrle in the development of a SPI component is the vast amount of different SPI devices, with different protocols that need to be supported. This means that each sensor might need a separate component....
- modifying the Encoder module like you ask. Will do this, but you have to set counter mode when you compile the firmware! This is something you thus cannot switch on or off at demand. From the number of counts you're on your own though...

**Reply by Peter-van-Tol** [reply]
I created a new issue (#110) and branch for this feature request. It will be up for testing tonight I think.

**Reply by Peter-van-Tol** [reply]
The work is done in #111. If you would be so kind to test it? When it is working, please let me know, then I merge it into `main`.

To install Litex-CNC for testing, you can run `pip install git+https://github.com/Peter-van-Tol/LiteX-CNC@110-feature-request-support-encoder-counter-mode`.

**Reply by Peter-van-Tol** [reply]
BTW: Did you mean an LM3**3**1 as a converter? I'm really curious on your schematics for the sensor.

**Reply by horcsab** [reply]
The schenmatic is not ready yet, but I want to do it this week.
I Yes, I meant LM331 or similar U/f converter. AD7740, AD654.
I used to use the LM331, but in a different circuit. It's been working fine ever since.

**Reply by Peter-van-Tol** [reply]
Let me know if you have any trouble in installing the code from the branch. I really like the idea behind the ADC through the frequency, because it eliminates the need for SPI or I2C at the moment (although, at one time I need to support it).

**Comment by Peter-van-Tol** [comment]
Is the proposed solution fulfilling your needs? As far as I can find the frequency of those converters is around 1 MHz tops. Given a servo-period of 1 kHz, this gives you a maximum resolution of 1000 pulses per servo-cycle, or an resolution of 10 bits. This gets worse with faster servo-threads or slower converters.

If you would like to increase the resolution, one has to count multiple servo periods . For example for 12 bits you have to count for roughly 4 ms (or 4 periods in my example), for 16 bits it is already 66 ms. In that case it would be more accurate to create a new component in LitexCNC for these chips:
- the component will provide an optional clock-out signal with a certain frequency (i.e. you don't have to abuse a PWM for this);
- the component will count 1,000,000 clock pulses out (number should be configurable) and determines how many were received back.
- when the total number of clock pulses have been received, the data register is updated with the new value and the cycle is restarted.

Such a component would be easy to build as an external component.

**Reply by hmnijp** [reply]
I looked at your commit. you write:

> The pin on the FPGA-card for Encoder B-signal. This pin is optional, ”
>         “when not set the encoder will be in `counter-mode`. In this mode the encoder will ”
>         “increment on the rising edge of the A-signal.

This means that the encoder counter can only increse.

I also saw in the description of mesa counter-mode that they use phase B for DIR (increase or decrease counter) maybe this will be useful too (capturing the stepgen signal for example, or non-quad encoders):

> (bit r/w) encoder.counter-mode
> 
> Set to False (the default) for Quadrature. Set to True for Step/Dir (in which case Step is on the A pin and Dir is on the B pin).

**Reply by Peter-van-Tol** [reply]
Thanks for this direction. I will add your recommendation to the feature request. 

At the moment I'm building a true high res counter, which can be used for determining the frequency of a signal. Will see what I can do on Wednesday.

**Comment by horcsab** [comment]
Hello Peter,

I took out the U/f converter again.
I was wondering if it would be possible to measure the period time with the FPGA? I think we would get much better resolution than with the encoder counting mode.
I don't know if there is a solution to handle this in Linuxcnc.
